From c3bd1ff2d9f67eb39c39b58bf3b3679d9a9a4894 Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Wed, 18 Oct 2023 14:57:14 +0530
Subject: [PATCH 3/4] hostapd: de-initialize/disable link BSS properly

Currently, if first link BSS of an interface is getting de-initialized/
disabled, then whole MLD is brought down. All other links are made
to stop beaconing and links are removed. And if, non-first link BSS
is de-initialized/disabled, nothing happens. Even beaconing is not
stopped which is wrong.

Fix the above by properly bringing down the intended link alone from the
interface. If only 1 link was there, then the interface will be removed.

And if first link BSS is removed, make the next link as the the new first
link of the BSS. Accordingly, cache the bss->ctx and drv->ctx as well.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 hostapd/main.c               |   1 +
 src/ap/ap_drv_ops.c          |  17 +++++
 src/ap/ap_drv_ops.h          |   2 +
 src/ap/hostapd.c             | 128 ++++++++++++++++++++++++-----------
 src/drivers/driver.h         |  27 ++++++++
 src/drivers/driver_nl80211.c |  84 +++++++++++++++++++++++
 src/drivers/driver_nl80211.h |   1 +
 7 files changed, 219 insertions(+), 41 deletions(-)

--- a/hostapd/main.c
+++ b/hostapd/main.c
@@ -181,6 +181,7 @@ static int hostapd_driver_init(struct ho
 
 	if (h_hapd) {
 		hapd->drv_priv = h_hapd->drv_priv;
+		hapd->interface_added = h_hapd->interface_added;
 
 		/*
 		 * All interfaces participating in the AP MLD would have
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -535,12 +535,29 @@ int hostapd_if_add(struct hostapd_data *
 }
 
 
+int hostapd_if_link_remove(struct hostapd_data *hapd, enum wpa_driver_if_type type,
+			   const char *ifname, u8 link_id)
+{
+	if (hapd->driver == NULL || hapd->drv_priv == NULL ||
+	    hapd->driver->if_link_remove == NULL)
+		return -1;
+
+	return hapd->driver->if_link_remove(hapd->drv_priv, type, ifname,
+					    hapd->mld_link_id);
+}
+
 int hostapd_if_remove(struct hostapd_data *hapd, enum wpa_driver_if_type type,
 		      const char *ifname)
 {
 	if (hapd->driver == NULL || hapd->drv_priv == NULL ||
 	    hapd->driver->if_remove == NULL)
 		return -1;
+
+#ifdef CONFIG_IEEE80211BE
+	if (hapd->conf->mld_ap)
+		return hostapd_if_link_remove(hapd, type, ifname, hapd->mld_link_id);
+#endif /* CONFIG_IEEE80211BE */
+
 	return hapd->driver->if_remove(hapd->drv_priv, type, ifname);
 }
 
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -61,6 +61,8 @@ int hostapd_if_add(struct hostapd_data *
 		   const char *bridge, int use_existing);
 int hostapd_if_remove(struct hostapd_data *hapd, enum wpa_driver_if_type type,
 		      const char *ifname);
+int hostapd_if_link_remove(struct hostapd_data *hapd, enum wpa_driver_if_type type,
+			   const char *ifname, u8 link_id);
 int hostapd_set_ieee8021x(struct hostapd_data *hapd,
 			  struct wpa_bss_params *params);
 int hostapd_get_seqnum(const char *ifname, struct hostapd_data *hapd,
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -477,28 +477,6 @@ static int hostapd_broadcast_wep_set(str
 
 #endif /* CONFIG_WEP */
 
-
-static void hostapd_clear_drv_priv(struct hostapd_data *hapd)
-{
-	unsigned int i, j;
-
-	for (i = 0; i < hapd->iface->interfaces->count && hapd->conf->mld_ap; i++) {
-		struct hostapd_iface *iface = hapd->iface->interfaces->iface[i];
-
-		if (!iface || hapd->iface == iface)
-			continue;
-
-		for (j = 0; j < iface->num_bss; j++) {
-			if (iface->bss && iface->bss[j] &&
-			    iface->bss[j]->mld_first_bss == hapd)
-				iface->bss[j]->drv_priv = NULL;
-		}
-	}
-
-	hapd->drv_priv = NULL;
-}
-
-
 void hostapd_free_hapd_data(struct hostapd_data *hapd)
 {
 	os_free(hapd->probereq_cb);
@@ -558,7 +536,7 @@ void hostapd_free_hapd_data(struct hosta
 			 * driver wrapper may have removed its internal instance
 			 * and hapd->drv_priv is not valid anymore.
 			 */
-			hostapd_clear_drv_priv(hapd);
+			hapd->drv_priv = NULL;
 		}
 	}
 
@@ -615,6 +593,104 @@ void hostapd_free_hapd_data(struct hosta
 #endif /* CONFIG_IEEE80211AX */
 }
 
+static void hostapd_update_first_link_bss(struct hostapd_data *hapd,
+					  struct hostapd_data *new_first_bss)
+{
+	struct hapd_interfaces *interfaces = hapd->iface->interfaces;
+	struct hostapd_iface *iface;
+	struct hostapd_data *bss;
+	struct hostapd_bss_config *conf;
+	size_t i, j;
+
+	for (i = 0; i < interfaces->count; i++) {
+		iface = interfaces->iface[i];
+		if (!iface || iface == hapd->iface)
+			continue;
+
+		for (j = 0; j < iface->num_bss; j++) {
+			bss = iface->bss[j];
+			if (!bss)
+				continue;
+
+			if (bss == new_first_bss)
+				continue;
+
+			conf = bss->conf;
+			if (!conf->mld_ap)
+				continue;
+
+			if (!hostapd_is_ml_partner(hapd, bss))
+				continue;
+
+			wpa_printf(MSG_DEBUG, "MLD %s: Link %d: Updated first_link BSS %d -> %d",
+				   conf->iface, bss->mld_link_id,
+				   hapd->mld_first_bss ? hapd->mld_first_bss->mld_link_id :
+				   			 hapd->mld_link_id,
+				   new_first_bss->mld_link_id);
+			bss->mld_first_bss = new_first_bss;
+		}
+	}
+}
+
+/* hostapd_bss_link_deinit - Per-BSS ML cleanup (deinitialization)
+ * @hapd: Pointer to BSS data
+ *
+ * This function is used to unlink the BSS from the partner BSSes.
+ * If the BSS being removed is the first link and other link exists,
+ * then it passes the data stored in it to the next link and make that
+ * link as the first BSS.
+ */
+static void hostapd_bss_link_deinit(struct hostapd_data *hapd)
+{
+#ifdef CONFIG_IEEE80211BE
+	if (hapd->conf->mld_ap && hapd->drv_priv &&
+	    !is_zero_ether_addr(hapd->mld_addr)) {
+		if (!hapd->mld_first_bss) {
+			struct hostapd_data *next, *link_bss;
+
+			/* Going to delete first BSS in MLD. Make the next entry as
+			 * first BSS and adjust first bss pointer of the rest all
+			 * link BSS
+			 */
+			if (dl_list_len(&hapd->links) >= 1) {
+				next = dl_list_entry(hapd->links.next,
+						     struct hostapd_data,
+						     links);
+
+				/* adjust first BSS pointer for all partner links
+				 * Can't use for_each_partner_bss() since a few
+				 * partner links could be in DISABLED state hence
+				 * unlinked from the list temproarily
+				 */
+				hostapd_update_first_link_bss(hapd, next);
+
+				/* copy mld_next_link_id value from old first BSS
+				 * to this new first BSS
+				 */
+				next->mld_next_link_id = hapd->mld_next_link_id;
+
+				/* make this as first BSS */
+				next->mld_first_bss = NULL;
+
+				/* make the old first BSS to point to this new
+				 * first BSS
+				 */
+				hapd->mld_first_bss = next;
+
+				wpa_printf(MSG_DEBUG, "MLD %s: Link %d: First link changed to %d",
+					   hapd->conf->iface, hapd->mld_link_id,
+					   next->mld_link_id);
+				wpa_printf(MSG_DEBUG, "MLD %s: First link changed %d -> %d",
+					   hapd->conf->iface, hapd->mld_link_id,
+					   next->mld_link_id);
+			}
+		}
+		wpa_printf(MSG_DEBUG, "MLD: Link %d unlinked from MLD interface %s",
+			   hapd->mld_link_id, hapd->conf->iface);
+		dl_list_del(&hapd->links);
+	}
+#endif
+}
 
 /**
  * hostapd_cleanup - Per-BSS cleanup (deinitialization)
@@ -1362,6 +1438,7 @@ static int hostapd_setup_bss(struct host
 					   hapd->mld_link_id, hapd->conf->iface);
 				dl_list_add_tail(&h_hapd->links, &hapd->links);
 				os_memcpy(hapd->mld_addr, h_hapd->mld_addr, ETH_ALEN);
+				hapd->interface_added = h_hapd->interface_added;
 				goto setup_mld;
 			}
 		}
@@ -1777,6 +1854,7 @@ static int start_ctrl_iface(struct hosta
 static void hostapd_no_ir_cleanup(struct hostapd_data *bss)
 {
 	hostapd_bss_deinit_no_free(bss);
+	hostapd_bss_link_deinit(bss);
 	hostapd_free_hapd_data(bss);
 	hostapd_cleanup_iface_partial(bss->iface);
 }
@@ -2512,6 +2590,7 @@ static int hostapd_setup_interface_compl
 			for (;;) {
 				hapd = iface->bss[j];
 				hostapd_bss_deinit_no_free(hapd);
+				hostapd_bss_link_deinit(hapd);
 				hostapd_free_hapd_data(hapd);
 				if (j == 0)
 					break;
@@ -2530,6 +2609,7 @@ static int hostapd_setup_interface_compl
 				for (;;) {
 					hapd = iface->bss[j];
 					hostapd_bss_deinit_no_free(hapd);
+					hostapd_bss_link_deinit(hapd);
 					hostapd_free_hapd_data(hapd);
 					if (j == 0)
 						break;
@@ -2840,10 +2920,8 @@ static void hostapd_bss_deinit(struct ho
 		hapd->rad_attr_db = NULL;
 	}
 #endif /* CONFIG_SQLITE */
-#ifdef CONFIG_IEEE80211BE
-	if (hapd->conf->mld_ap)
-		dl_list_del(&hapd->links);
-#endif
+
+	hostapd_bss_link_deinit(hapd);
 	hostapd_cleanup(hapd);
 }
 
@@ -3115,6 +3193,40 @@ hostapd_interface_init_bss(struct hapd_i
 	return iface;
 }
 
+static void hostapd_cleanup_driver(const struct wpa_driver_ops *driver,
+				   void *drv_priv,
+				   struct hostapd_iface *iface)
+{
+	bool ap_mld = false;
+
+	if (!driver || !driver->hapd_deinit || !drv_priv)
+		return;
+
+#ifdef CONFIG_IEEE80211BE
+	ap_mld = !!iface->bss[0]->conf->mld_ap;
+#endif /* CONFIG_IEEE80211BE */
+
+	/* In case of non-ML operation, de-init. But, if ML operation exist, then
+	 * even if thats the last BSS in the interface, the driver (drv) could
+	 * be in use for a different MLD. Hence, need to check if drv is still
+	 * being used by some other bss before de-initiallizing
+	 */
+	if (!ap_mld)
+		driver->hapd_deinit(drv_priv);
+	else if (!iface->bss[0]->mld_first_bss &&
+		 driver->get_drv_shared_status &&
+		 !driver->get_drv_shared_status(drv_priv, iface->bss[0])) {
+		driver->hapd_deinit(drv_priv);
+	} else if (hostapd_if_link_remove(iface->bss[0],
+					 WPA_IF_AP_BSS,
+					 iface->bss[0]->conf->iface,
+					 iface->bss[0]->mld_link_id)) {
+		wpa_printf(MSG_WARNING, "Failed to remove BSS interface %s",
+			   iface->bss[0]->conf->iface);
+	}
+
+	iface->bss[0]->drv_priv = NULL;
+}
 
 void hostapd_interface_deinit_free(struct hostapd_iface *iface)
 {
@@ -3133,11 +3245,7 @@ void hostapd_interface_deinit_free(struc
 	hostapd_interface_deinit(iface);
 	wpa_printf(MSG_DEBUG, "%s: driver=%p drv_priv=%p -> hapd_deinit",
 		   __func__, driver, drv_priv);
-	if (driver && driver->hapd_deinit && drv_priv) {
-		if (!iface->bss[0]->mld_first_bss)
-			driver->hapd_deinit(drv_priv);
-		hostapd_clear_drv_priv(iface->bss[0]);
-	}
+	hostapd_cleanup_driver(driver, drv_priv, iface);
 	hostapd_interface_free(iface);
 }
 
@@ -3176,19 +3284,22 @@ static void hostapd_deinit_driver(const
 
 	wpa_printf(MSG_DEBUG, "%s: driver=%p drv_priv=%p -> hapd_deinit",
 		   __func__, driver, drv_priv);
+
+	hostapd_cleanup_driver(driver, drv_priv, hapd_iface);
+
 	if (driver && driver->hapd_deinit && drv_priv) {
-		if (!hapd_iface->bss[0]->mld_first_bss)
-			driver->hapd_deinit(drv_priv);
 		for (j = 0; j < hapd_iface->num_bss; j++) {
 			wpa_printf(MSG_DEBUG, "%s:bss[%d]->drv_priv=%p",
 				   __func__, (int) j,
 				   hapd_iface->bss[j]->drv_priv);
-			if (hapd_iface->bss[j]->drv_priv == drv_priv) {
-				hostapd_clear_drv_priv(hapd_iface->bss[j]);
-				hapd_iface->extended_capa = NULL;
-				hapd_iface->extended_capa_mask = NULL;
-				hapd_iface->extended_capa_len = 0;
-			}
+
+			if (hapd_iface->bss[j]->drv_priv != drv_priv)
+				continue;
+
+			hapd_iface->bss[j]->drv_priv = NULL;
+			hapd_iface->extended_capa = NULL;
+			hapd_iface->extended_capa_mask = NULL;
+			hapd_iface->extended_capa_len = 0;
 		}
 	}
 }
@@ -3307,6 +3418,7 @@ int hostapd_disable_iface(struct hostapd
 	for (j = 0; j < hapd_iface->num_bss; j++) {
 		struct hostapd_data *hapd = hapd_iface->bss[j];
 		hostapd_bss_deinit_no_free(hapd);
+		hostapd_bss_link_deinit(hapd);
 		hostapd_free_hapd_data(hapd);
 	}
 
@@ -3505,6 +3617,7 @@ int hostapd_add_iface(struct hapd_interf
 			if (start_ctrl_iface_bss(hapd) < 0 ||
 			    (hapd_iface->state == HAPD_IFACE_ENABLED &&
 			     hostapd_setup_bss(hapd, -1, true))) {
+				hostapd_bss_link_deinit(hapd);
 				hostapd_cleanup(hapd);
 				hapd_iface->bss[hapd_iface->num_bss - 1] = NULL;
 				hapd_iface->conf->num_bss--;
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -5184,6 +5184,33 @@ struct wpa_driver_ops {
 			      const u8 *match, size_t match_len,
 			      bool multicast);
 #endif /* CONFIG_TESTING_OPTIONS */
+
+#ifdef CONFIG_IEEE80211BE
+	/**
+	 * if_link_remove - Remove a link alone from virtual interface
+	 * @priv: Private driver interface data
+	 * @type: Interface type
+	 * @ifname: Interface name of the virtual interface from where link is
+	 *	    to be removed
+	 * @link_id: Valid link ID to remove
+	 * Returns: 0 on success, -1 on failure
+	 */
+	int (*if_link_remove)(void *priv, enum wpa_driver_if_type type,
+			      const char *ifname, int link_id);
+
+	/**
+	 * get_drv_shared_status - Get shared status of driver interface
+	 * @priv: private driver interface data from init()
+	 * @bss_ctx: BSS context for %WPA_IF_AP_BSS interfaces
+	 *
+	 * Checks whether driver interface is being used by other partner BSS(s)
+	 * or not. This is used to decide whether driver interface needs to be
+	 * deinitilized when one iface is getting deinitialized.
+	 *
+	 * Returns: True if it is being used or else False.
+	 */
+	bool (*get_drv_shared_status)(void *priv, void *bss_ctx);
+#endif /* CONFIG_IEEE80211BE */
 };
 
 /**
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -4121,6 +4121,19 @@ int wpa_driver_nl80211_authenticate_retr
 	return wpa_driver_nl80211_authenticate(bss, &params);
 }
 
+int nl80211_valid_link(struct i802_bss *bss, s8 link_id)
+{
+	unsigned int i;
+
+	for (i = 0; i < bss->n_links; i++) {
+		if (bss->links[i].link_id != link_id)
+			continue;
+
+		return 1;
+	}
+
+	return 0;
+}
 
 struct i802_link * nl80211_get_link(struct i802_bss *bss, s8 link_id)
 {
@@ -8965,6 +8978,37 @@ static int wpa_driver_nl80211_if_remove(
 	return 0;
 }
 
+static int wpa_driver_nl80211_if_link_remove(struct i802_bss *bss,
+					     enum wpa_driver_if_type type,
+					     const char *ifname,
+					     int link_id)
+{
+	struct wpa_driver_nl80211_data *drv = bss->drv;
+	struct i802_link *link;
+	int num_links = bss->n_links;
+
+	wpa_printf(MSG_DEBUG, "nl80211: %s(type=%d ifname=%s num_links=%d) link_id=%d",
+		   __func__, type, ifname, num_links, link_id);
+
+	link = nl80211_get_link(bss, link_id);
+
+	wpa_printf(MSG_DEBUG, "nl80211: Teardown AP(%s) link %d", bss->ifname,
+		   link_id);
+	wpa_driver_nl80211_del_beacon(bss, link);
+	nl80211_remove_link(bss, link_id);
+
+	bss->ctx = bss->links[0].ctx;
+
+	if (drv->first_bss == bss && num_links > 1)
+		drv->ctx = bss->ctx;
+
+	if (num_links == 1) {
+		wpa_printf(MSG_DEBUG, "nl80211: Only 1 link was there hence remove interface");
+		return wpa_driver_nl80211_if_remove(bss, type, ifname);
+	}
+
+	return 0;
+}
 
 static int cookie_handler(struct nl_msg *msg, void *arg)
 {
@@ -10608,6 +10652,22 @@ static int driver_nl80211_if_remove(void
 	return wpa_driver_nl80211_if_remove(bss, type, ifname);
 }
 
+static int driver_nl80211_if_link_remove(void *priv, enum wpa_driver_if_type type,
+					 const char *ifname, int link_id)
+{
+	struct i802_bss *bss = priv;
+
+	if (link_id < 0 || link_id >= MAX_NUM_MLD_LINKS)
+		return -1;
+
+	if (type != WPA_IF_AP_BSS)
+		return -1;
+
+	if (!nl80211_valid_link(bss, link_id))
+		return -1;
+
+	return wpa_driver_nl80211_if_link_remove(bss, type, ifname, link_id);
+}
 
 static int driver_nl80211_send_mlme(void *priv, const u8 *data,
 				    size_t data_len, int noack,
@@ -13800,6 +13860,29 @@ fail:
 	return ret;
 }
 
+bool wpa_driver_get_shared_status(void *priv, void *bss_ctx)
+{
+	struct i802_bss *bss = priv;
+	struct wpa_driver_nl80211_data *drv = bss->drv;
+
+	for (bss = drv->first_bss; bss; bss = bss->next) {
+		/* more than one links means some one is still using */
+		if (bss->n_links > 1)
+			return true;
+
+		/* if its 1 then it could be SLO or invalid entry. Need
+		 * to check the link ID. But need to check only if the link
+		 * BSS context is not same as self since self can exist and
+		 * will be removed when deinitializing the driver
+		 */
+		if (bss->flink->ctx != bss_ctx &&
+		    bss->flink->link_id != NL80211_DRV_LINK_ID_NA)
+			return true;
+	}
+
+	return false;
+}
+
 const struct wpa_driver_ops wpa_driver_nl80211_ops = {
 	.name = "nl80211",
 	.desc = "Linux nl80211/cfg80211",
@@ -13954,4 +14037,8 @@ const struct wpa_driver_ops wpa_driver_n
 	.radio_disable = testing_nl80211_radio_disable,
 #endif /* CONFIG_TESTING_OPTIONS */
 	.set_6gclient_type = nl80211_set_6gclient_type,
+#ifdef CONFIG_IEEE80211BE
+	.if_link_remove = driver_nl80211_if_link_remove,
+	.get_drv_shared_status = wpa_driver_get_shared_status,
+#endif /* CONFIG_IEEE80211BE */
 };
--- a/src/drivers/driver_nl80211.h
+++ b/src/drivers/driver_nl80211.h
@@ -326,6 +326,7 @@ const char * nl80211_iftype_str(enum nl8
 
 void nl80211_restore_ap_mode(struct i802_bss *bss);
 struct i802_link * nl80211_get_link(struct i802_bss *bss, s8 link_id);
+int nl80211_valid_link(struct i802_bss *bss, s8 link_id);
 
 #ifdef ANDROID
 int android_nl_socket_set_nonblocking(struct nl_sock *handle);
